### ISSUE TITLE
`cargo test --all-targets` does not run doc tests

### ISSUE THREAD

**Issue by thomaseizinger** [rust-lang/cargo#6669]
As per title `cargo test --all-targets` does not run the doc tests on a crate. This is very surprising behaviour. I found [this](https://github.com/rust-lang/cargo/pull/4592#issuecomment-338409476) and it looks like running doc tests is done by `rustdoc` which is apparently not invoked when doing `--all-targets`.

**Comment by dwijnand** [comment]
An attempt to fix this was made in #6671, but there's a bit of a mismatch between `cargo test` and doc tests that makes the issue complicated, so the PR didn't land.

**Comment by thomaseizinger** [comment]
Maybe it is worth to add some context to this issue.

We ran into this because we use `cargo test --all-targets` on Travis and usually `cargo test` locally.
A not-compiling doc test sneaked into a last-minute update of a PR but the build was fine since we use `--all-targets` on Travis.
Later on, someone complained that we have a broken `master` because of a non-compiling doc test.

I read some of the discussion in the PR and TBH, the whole `--doc` is a mode would be SUPER confusing.
As a user, I expect `cargo test` to test LESS than `cargo test --all-targets` since the latter has `all` in the name.
In addition, the documentation (`--help`) lists doc tests next to all the other targets.

**Comment by glts** [comment]
This means the command to run absolutely all tests everywhere is `cargo test --workspace`, not `cargo test --workspace --all-targets`, right?

**Comment by ehuss** [comment]
> This means the command to run absolutely all tests everywhere is `cargo test --workspace`, not `cargo test --workspace --all-targets`, right?

It depends on the structure of your packages, but in general there is currently no single command to "test everything".   `cargo test` is intended to be the usual command.  If you also want to test benchmarks and examples, you can either mark them with `test=true` in `Cargo.toml`, or run two commands (`cargo test --all-targets` and `cargo test --doc`).

**Comment by Veetaha** [comment]
Oh god, I think we've been not running doc tests on our CI for a couple of years due to this...

**Comment by ehuss** [comment]
Just for some clarification, cargo has a model where there is a distinction between a *target* and a *mode*. Doctests are a *mode*, and that is why `--all-targets` by itself doesn't run them. I understand that can be confusing, isn't obvious from the documentation, and not the mental model some people have. We have talked in the past about making that more explicit (which in part would solve this issue by offering new switches to do both "run tests" and "run doctests"). I can't find that discussion offhand, but I don't think it was very deep anyway (just acknowledging that the CLI is limited and there were some ideas around a `--mode` flag). Perhaps that's something the testing team could tackle 😉 .

**Comment by epage** [comment]
> doctests are a mode, and that is why --all-targets by itself doesn't run them. I understand that can be confusing, isn't obvious from the documentation

Documentation is hard
- In [Documentation tests](https://doc.rust-lang.org/cargo/commands/cargo-test.html#documentation-tests), we say "Different from normal test targets", implying it is a target
- In [Target selection](https://doc.rust-lang.org/cargo/commands/cargo-test.html#target-selection), we talk abut doctests being a target

**Comment by obi1kenobi** [comment]
The [`--all-targets` portion of Target selection](https://doc.rust-lang.org/cargo/commands/cargo-test.html#target-selection) also needs an edit. I attempted to do that in #12423, but the current structure of the docs makes it difficult to make a carve-out that adds the additional `--all-targets` clarification for `cargo test` only, without applying it to `--all-targets` for other cargo commands that have nothing to do with doctests.

**Comment by weihanglo** [comment]
I'd like to copy over [some proposals from ehuss](https://github.com/rust-lang/cargo/pull/6671#issuecomment-464202979) that was buried in a closed PR. Fundamentally this is more a UX issue than performance or technical issue.

---
Method A: `--doc` is a mode flag.
- `--doc` will doctest all doctestable targets.
- `--doc <target flags>` will only doctest the given targets.
- `--doc --all-targets` would only doctest all targets.
- `--all-targets` would only rustc test all targets.  (Or, maybe it could do both tests and doctests?)
- `<target flags>` will only rustc test the given targets.

Drawbacks here: I'm not sure if this is clear.  If `--all-targets` only does rustc tests, then you would need two commands (`cargo test --all-targets` and `cargo test --doc --all-targets`) to do everything.  If it does both, then it is a special case that people need to learn about.  You wouldn't be able to say "test and doctest all examples", unless all plurality flags like `--examples` are special, in which case I think it would be confusing if `--example foo` didn't run doctests (but that seems undesirable to me).

Method B: `--mode=...` can specify the types of tests that are run.
- `--mode=test,doctest <target flags>` will test *and* doctest the given targets.
- No `--mode` means rustc test only.

Drawbacks is that this adds another flag, which is very undesirable.

Method C: `--doc` could take targets as a parameter
- `--doc --lib` doctests all doctestable targets, and rustc tests just the library.
- `--doc=lib` doctests just the lib.
- `--doc=lib,example=foo,bins`  doctests lib, example foo, and all bins.
- `--doc=all-targets --all-targets` does everything

Drawbacks is that I think the first case is confusing, and the last one is awkward.

---

**Comment by est31** [comment]
Yeah it would be great if there was a single command to run all tests, doctests, everything.

having a unified command is a blocker for re-attempting #8437

**Comment by thomaseizinger** [comment]
> Just for some clarification, cargo has a model where there is a distinction between a _target_ and a _mode_. Doctests are a _mode_, and that is why `--all-targets` by itself doesn't run them. I understand that can be confusing, isn't obvious from the documentation, and not the mental model some people have. We have talked in the past about making that more explicit (which in part would solve this issue by offering new switches to do both "run tests" and "run doctests"). I can't find that discussion offhand, but I don't think it was very deep anyway (just acknowledging that the CLI is limited and there were some ideas around a `--mode` flag). Perhaps that's something the testing team could tackle 😉 .

I don't know anything about the internals of `cargo` and I've been using Rust for close to 6 years now in a full-time setting. This is the first time I am hearing about `cargo` having a mode (other than the "release mode" toggled by `--release`).

Can we just remove the `--doc` flag and always include doc tests in every target that is selected, i.e. `--lib`, `--tests`, `--examples` etc. The usecase of only running doc tests could be served by a new flag of `cargo doc`: `cargo doc --tests` although I'd probably play it conservative and not add that one to begin with. It seems odd wanting to run _only_ the doc tests. If doc tests are just more tests that get compiled into the binary, the could be selected via the regular filtering of test names. Perhaps they should all be under a dedicated `doc` module? Something like `cargo test mycrate::doc::my_function_name_that_has_doc_test`.

**Comment by est31** [comment]
> Can we just remove the `--doc` flag and always include doc tests in every target that is selected

doctests are not unit tests, they are integration tests. In other words, they require a separate rustc process to be spawned for each snippet. On top of that, you have the rustdoc process (which isn't *that* neccessary, one could move the logic to extract the snippets to a place shared by rustc and rustdoc).

**Comment by joshtriplett** [comment]
In today's @rust-lang/cargo meeting, we agreed to try the experiment of adding an `--all-targets-including-doc` unstable option, initially to `cargo test` and then to other commands like `cargo build` and `cargo check`. We don't yet know what we should call it, but we can experiment with implementation of it.

**Comment by epage** [comment]
Also, when we do `build`, `check`, and `fix` (and suggest for `clippy` to do similar), we might want to add `--doc` flags to them (#6424).

As for naming, one idea is to explore disambiguating the term ["target"](https://doc.rust-lang.org/cargo/appendix/glossary.html#target) in cargo.  If we rename "build target", then the `--all-<something>` could just have the new meaning.

We lean towards a new flag because people likely have broken doctests (as evidenced by this thread).  While they would want to discover and fix them, they likely want to do it on their own timetable and not be forced to by cargo.

**Comment by joshtriplett** [comment]
Note that in a future edition we could potentially make `--all-targets` default to this, which would make this an interim option for current editions.

**Comment by djc** [comment]
Hmm, I wonder if we can have a project-level switch (maybe in the `.cargo/config`) for this? `cargo c --all-features --all-targets` is already pretty verbose and I'm not sure making it more onerous to do the right thing is the best way forward. So for projects that are aware of this and want to eagerly fix the doctests breakage, it would be great if we can opt-in to making `--all-targets` do (arguably) the right thing without waiting until 2027.

**Comment by epage** [comment]
I think we're leaning towards [renaming "build `target`"](https://internals.rust-lang.org/t/renaming-cargo-target/21585) and having the new name for `--all-targets` include doctests.  We would then not need to wait for a new edition.

**Comment by weihanglo** [comment]
Just note that there are designs needing to figure out, for the interaction between doctesting everything and a subset of doctest-able lib/example/bench/test/bin. See <https://github.com/rust-lang/cargo/issues/6669#issuecomment-1850761204>

**Comment by thomaseizinger** [comment]
> Hmm, I wonder if we can have a project-level switch (maybe in the `.cargo/config`) for this?

You could maybe solve that with an alias that "does the right thing" whatever that means for you?

**Comment by orium** [comment]
> I think we're leaning towards renaming "build target" and having the new name for --all-targets include doctests. We would then not need to wait for a new edition.

@epage The idea would still be to default to the "`--all-targets` includes docs" in edition 2024, is that right? The new edition is just a few months away, so I would be happy just waiting for that instead of having an `--all-targets-with-doc`.

**Comment by epage** [comment]
It is too late to introduce 2024 Edition changes. If a solution needs an Edition, it'd be 2027.

**Comment by orium** [comment]
Ah, well, it that case it makes sense to have a new flag for "all targets and docs" 😄 .
